### PR TITLE
feat(efb): flyPad battery

### LINKED ISSUE #5615: FlyPad battery life
### Aircraft Version

Stable

### Description

If you look up at the top right corner of the Flypad, you see the battery meter and how it is always charging. I think it should be that while AC power isn't being supplied to the plane (Both APU and exterior power are off) the flypad doesn't charge and the battery starts to drain, like how the flight instruments aren't powered.

### References (optional)

_No response_

### Additional info (optional)

Similar to how the flight instruments are not powered when neither the APU or EXT PWR is avail, the same should be for the FlyPad's charging supply.

### Discord Username (optional)

_No response_

### PR DESCRIPTION
**Pull request by schmidHolger**
Fixes #5615

## Summary of Changes
Manage battery level for EFB fly pad.

Battery is charged as long as DC2 bus is powerd. If not, battery discharges for 240min using a very simple linear function. Charging takes 180min from empty to full - again using a linear function model.

In case the battery is empty, flypad shuts down completly and needs to be charaged again to be functional again. 

### (1) Batttery full, DC2 bus powered
01_EfbBattery_Full-charged

### (2) Battery discharged, DC2 bus not powered
02_EfbBattery-discharge

### (3) Battery at warning level
03_EfbBattery-warning-discharge

### (4) Battery empty, flypad not operational
04_EfbBattery-empty

### (5) Flypad not yet operational (close to empty) but charged
05_EfbBattery-empty-charge

### (6) Flypad operational, still being charged
06_EfbBattery-charge

## References
n/a

Battery life parameters taken from own experience but not from a real aircraft ;-).

## Additional context

The battery model is based on ABSOLUTE TIME, therefore changing the time in the sim should also change the battery level accordingly. 

Discord username (if different from GitHub): `geholger [Z+1]#9751`

## Testing instructions

1. Start EFB
2. Disconnect DC2 bus --> fly pad should discharge and battery level / icon should change over time (see (1))
3. Connect DC2 bus --> fly pad should discharge (see (6), battery level should increase
4. Disconect DC2 bus and wait until battey is empty (you might want to change the sim time)
--> at <8%, battery warning should be indicated (see (3))
--> at 0%,  battery level, fly pad should shut down (see (4))
5. Connect DC2 bus
--> Fly pad should remain inoperative for a while but "charging" should be shown (see (5))
--> After a while, fly pad should start up and be functional again (see (6))

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by tracernz** [comment]
How do you recover from a failure that takes out DC2 if the flyPad battery is flat?

**Comment by schmidHolger** [comment]
If the flypad is empty and DC2 has a failure than the efb stays on empty. What would you do in real life? Is there a way to hook up the fly pad to another bus?

**Comment by AviationSpotter** [comment]
Like real tablets, when getting below say 20%, a "low battery" message/warning should pop up.

**Comment by schmidHolger** [comment]
> Like real tablets, when getting below say 20%, a "low battery" message/warning should pop up.

To further drive that idea, we would need to add some powersaving mode to the efb (e.g. going dark after 5min after no input with a reduced battery consumption) but that I would see in a second step. With that available, it would make sense to have a pop up like described with the option to enter a power saving mode or just ack the low battery state,

As this would affect all EFB inputs and "touch" events everywhere on the EFB to reset the timer before going into this screnn-off power saving mode, I would like to do that in a seperate PR and then add the pop up with the option to enter power saving mode or not.

**Comment by tracernz** [comment]
> If the flypad is empty and DC2 has a failure than the efb stays on empty. What would you do in real life? Is there a way to hook up the fly pad to another bus?

In real life you'd get an engineer to fix it, but in the sim the EFB stands in... if it's flat I guess you have to quit. :P

**Comment by bouveng** [comment]
For testning: 🤓 

**TESTING OF CIRCUITS AND BUSES**
- To unpower AC BUS 1, push GEN 1 and BUS TIE to OFF.
- To unpower AC BUS 2, push GEN 2 and BUS TIE to OFF.
- To unpower AC ESS, push ESS FEED button to ALTN, BUS TIE to OFF and GEN 2 to OFF. This will also unpower AC BUS 2.
 - To unpower DC 1, DC 2, DC BAT, go to failures in the EFB and fail all TRs.

File: src/instruments/src/EFB/StatusBar/StatusBar.tsx [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
 
-export const StatusBar = () => {
+import { BatteryStatus } from './BatteryStatus';
+
+type StatusBarProps = {
+    batteryLevel: number;
+    isCharging: boolean;
+};
+
+export const StatusBar = (props: StatusBarProps) => {
```
Comment (ErickSharp) [inline src/instruments/src/EFB/StatusBar/StatusBar.tsx:14]: **suggestion**: Destructure here

File: src/instruments/src/EFB/StatusBar/BatteryStatus.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+const BATTERY_LEVEL_0 = 13;
+const BATTERY_LEVEL_1 = 37;
+const BATTERY_LEVEL_2 = 62;
+const BATTERY_LEVEL_3 = 87;
+const BATTERY_ICON_SIZE = 28;
+
+export const BatteryStatus = (props: BatteryStatusProps) => {
+    const { batteryLevel, isCharging } = props;
+    const batteryStatusIcon = () => {
+        if (isCharging === true) {
```
Comment (ErickSharp) [inline src/instruments/src/EFB/StatusBar/BatteryStatus.tsx:19]: No need for usage of comparison operator, replace with `if(isCharging)`

**Review by ErickSharp — CHANGES_REQUESTED** [review]

File: src/instruments/src/EFB/Efb.tsx [hunk 3]
```diff
@@ -44,11 +48,23 @@ const ScreenLoading = () => (
     </div>
 );
 
+const ScreenEmpty = (props: { charging: boolean }) => (
+    <div className="flex justify-center items-center w-screen h-screen bg-theme-statusbar">
+        {props.charging === false && (
```
[the remark this answers is not in the document]
Comment (schmidHolger) [reply src/instruments/src/EFB/Efb.tsx:53]: ok

File: src/instruments/src/EFB/Efb.tsx [hunk 4]
```diff
@@ -76,6 +100,43 @@ const Efb = () => {
     const [simbriefUserId] = usePersistentProperty('CONFIG_SIMBRIEF_USERID');
     const [autoSimbriefImport] = usePersistentProperty('CONFIG_AUTO_SIMBRIEF_IMPORT');
 
+    const [dc2BusIsPowered] = useSimVar('L:A32NX_ELEC_DC_2_BUS_IS_POWERED', 'bool');
+    const [batteryLevel, setBatteryLevel] = useState<BatteryLevel>({ level: 100, lastChangeTimestamp: absoluteTime, isCharging: dc2BusIsPowered });
+
+    useEffect(() => {
+        setBatteryLevel((oldLevel:BatteryLevel) => {
```
[the remark this answers is not in the document]
Comment (schmidHolger) [reply src/instruments/src/EFB/Efb.tsx:107]: ok

File: src/instruments/src/EFB/StatusBar/BatteryStatus.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+const BATTERY_LEVEL_WARNING = 8;
+const BATTERY_LEVEL_0 = 13;
+const BATTERY_LEVEL_1 = 37;
+const BATTERY_LEVEL_2 = 62;
+const BATTERY_LEVEL_3 = 87;
+const BATTERY_ICON_SIZE = 28;
+
+export const BatteryStatus = (props: BatteryStatusProps) => {
+    const { batteryLevel, isCharging } = props;
+    const batteryStatusIcon = () => {
```
[the remark this answers is not in the document]
Comment (schmidHolger) [reply src/instruments/src/EFB/StatusBar/BatteryStatus.tsx:18]: Yea - much better :-)

**Comment by ErickSharp** [comment]
Needs rebase

**Comment by schmidHolger** [comment]
> Needs rebase

Done.

**Review by ErickSharp — APPROVED** [review]
😁

**Comment by ShadowFrost3** [comment]
Quality Assurance Tester Report


Object of testing: #6392
Tier of Testing:1
Date: 01/01/2022

Testing Process: EFB does not turn on 
Testing Results: Not Passed

**Comment by schmidHolger** [comment]
> Quality Assurance Tester Report
> 
> Object of testing: #6392 Tier of Testing:1 Date: 01/01/2022
> 
> Testing Process: EFB does not turn on Testing Results: Not Passed

Strange - When using the downloaded files for this PR from github, the EFB is not starting up as described by @ShadowFrost3. Building from source locally and starting it up in the sim works as expected. Is this a build issue on github?

Brief analysis:
If I download the build from github, the folder `EFB` in `flybywire-aircraft-a320-neo\html_ui\Pages\VCockpit\Instruments\A32NX` seems to be missing. If I do a local build on my machine, the `EFB` folder exists. Any ideas what's going on here?

**Comment by lukecologne** [comment]
There are build errors on the EFB: 
```
✖ EFB
        Error when using sourcemap for reporting an error: Can't resolve original location of error.
        The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
        Error when using sourcemap for reporting an error: Can't resolve original location of error.
        The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
        Error when using sourcemap for reporting an error: Can't resolve original location of error.
        The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
        Error when using sourcemap for reporting an error: Can't resolve original location of error.
        The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
        Error when using sourcemap for reporting an error: Can't resolve original location of error.
        The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
        Error when using sourcemap for reporting an error: Can't resolve original location of error.
        The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
        file:///external/node_modules/rollup/dist/es/shared/rollup.js:10359
                base = Object.assign(new Error(base.message), base);
                                     ^
        
        Error: Could not resolve './Store/features/simbrief' from instruments/src/EFB/Efb.tsx
            at error (file:///external/node_modules/rollup/dist/es/shared/rollup.js:10359:30)
            at ModuleLoader.handleResolveId (file:///external/node_modules/rollup/dist/es/shared/rollup.js:22249:24)
            at file:///external/node_modules/rollup/dist/es/shared/rollup.js:22228:26 {
          code: 'UNRESOLVED_IMPORT',
```

**Comment by schmidHolger** [comment]
Build error on github fixed.

**Comment by lukecologne** [comment]
There are still errors:  
```
Error: Could not resolve './features/simbrief' from instruments/src/EFB/Store/store.ts
            at error (file:///external/node_modules/rollup/dist/es/shared/rollup.js:10359:30)
            at ModuleLoader.handleResolveId (file:///external/node_modules/rollup/dist/es/shared/rollup.js:22249:24)
            at file:///external/node_modules/rollup/dist/es/shared/rollup.js:22228:26 {
          code: 'UNRESOLVED_IMPORT',
```

**Comment by schmidHolger** [comment]
fixed two more import issues - unfortunatly they only appear in gitlab builds :-(

**Comment by lukecologne** [comment]
> fixed two more import issues - unfortunatly they only appear in gitlab builds :-(

Looks like it works now

**Comment by FlightSimmer7700** [comment]
Should the battery be discharging at the same rate even when the flypad is off?

**Comment by ErickSharp** [comment]
> Should the battery be discharging at the same rate even when the flypad is off?

We might want to consider it 0 since its almost negligible or we should see if we can pull up some data on average discharge rate for powered off devices!

**Comment by schmidHolger** [comment]
Currently, it's discharging at the same rate regardless of the EFB state. I could change it to e.g. 3 days if it's off but this would be close to 0.

**Comment by jcalor** [comment]
Quality Assurance Tester Report

Discord          : Calo#6816
Object of testing: #6392 
Tier of Testing  : 1
Date             : 03/01/2022

Testing Process: 

Start EFB - checked
Disconnect DC2 bus --> fly pad should discharge and battery level / icon should change over time (see (1)) - checked
Connect DC2 bus --> fly pad should discharge (see (6), battery level should increase - checked
Disconect DC2 bus and wait until battey is empty (you might want to change the sim time) - checked
--> at <8%, battery warning should be indicated (see (3)) - checked
--> at 0%, battery level, fly pad should shut down (see (4)) - checked
Connect DC2 bus
--> Fly pad should remain inoperative for a while but "charging" should be shown (see (5)) - checked
--> After a while, fly pad should start up and be functional again (see (6)) - checked

Negatives: N/A

Testing Results: Passed

Conclusions: flyPad v3 looks good! Worked nicely changing sim time. Good job!

**Comment by FlightSimmer7700** [comment]
Should the battery be discharging at the same rate even when the flypad is off?

> Currently, it's discharging at the same rate regardless of the EFB state. I could change it to e.g. 3 days if it's off but this would be close to 0.

Yeah probably would be better if it discharges at a slower rate while off. Maybe charging at a faster rate when off too?

**Comment by schmidHolger** [comment]
What about leaving this PR as is, getting it into the flypadosv3 branch and then adding a bug for the different discharge rates if EFB is off? I will setup  new PR for solving this issue then.

**Comment by mehoygg** [comment]
Quality Assurance Tester/Trainee Report

Discord          : mehoygg#1242
Object of testing: #6392
Tier of Testing  : 1
Date             : 03/01/22

Testing Process: 
Spawned in C/D at LAX.  
Started power on plane and turned on EFB.

Once EFB was powered, I looked for the charging icon on the battery, and it was 100% + Green.
I disconnected DC2B and verified to check that the EFB Percentage was not green, as in charging.
EFB started to discharge and I did speed up sim time to lower battery levels faster.  
Once low, I reconnected DC2B and at low battery percentage (red), and green charging symbol came up. 

I disconnected DC2B and let FlyPad hit 0% and it did shut off and showed the battery icon in red.
I waited a few mins and then reconnected DC2B and after a while let it charge.  Charge icon showed up on screen. Took about a minute or two to get it charged enough to turn on.  

Negatives:       
N/A
Testing Results: 
Passed

Conclusions: 
Seems really cool.  Being able to let it die/charge the flypad is a neat ft. 

Media:       
N/A

**Comment by D4rknessCodez** [comment]
Quality Assurance Trainee Report

Discord: Darkness🍁#7734
Object of testing: #6392
Tier of Testing: (1/2)
Date: (03/01/2022)

Testing Process: 
- Started the EFB once the plane had power
- After booting I disconnected the DC bus which led to the charge diminishing.
- After this, I reconnected the DC bus which led to the charge interesting.
- I disconnected the DC bus once again. When the battery was <8%, the battery warning icon had been indicated and when the battery was at 0%, the EFB had shut down and had a recharged battery symbol.
- I reconnected the DC bus and the charging symbol had shown up, once some time had passed by the EFB has relaunched and was fully operational and had been charging.

Negatives: N/A

Testing Results: 
Passed

Conclusions: 
Everything is working as intended and I had no performance issues whatsoever. FlypadOS v3 is looking sweet, Keep up the good work!

**Pull request merged**